### DISCUSSION TITLE
Quarkus 3.0.0.CR2 / Hibernate ORM 6 / Multitenancy DISCRIMINATOR Strategy

### DISCUSSION THREAD

**Discussion by humcqc** [quarkusio/quarkus#32686] [Q&A] [answered]
Hello,

I would like to use Discriminator strategy in Quarkus.
It was not available in 2.X due to hibernate version. So I used a specific filter and interceptor to implement this strategy.
Now that hibernate has been upgraded I've tried it.

But I didn't find a way to configure the Hibernate CurrentTenantIdentifierResolver.
Without CurrentTenantIdentifierResolver property, I have this issue:
```
2023-04-18 17:21:24,566 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-0) HTTP Request to /api/v1/condominiums failed, error id: 23a9a92b-2494-421f-8b9f-884589308e83-1: org.hibernate.HibernateException: SessionFactory configured for multi-tenancy, but no tenant identifier specified
	at org.hibernate.internal.SessionImpl.setUpMultitenancy(SessionImpl.java:276)
	at org.hibernate.internal.SessionImpl.<init>(SessionImpl.java:259)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.<init>(ReactiveSessionImpl.java:168)
	at org.hibernate.reactive.mutiny.impl.MutinySessionFactoryImpl.lambda$openSession$1(MutinySessionFactoryImpl.java:112)
```
It seems that the [FastBootEntityManagerFactoryBuilder](https://github.com/quarkusio/quarkus/blob/3.0/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootEntityManagerFactoryBuilder.java#L192) does not configure the CurrentTenantIdentifierResolver in case we just use @TenantId annotation without other properties.

Is there some documentation ? or is it something not yet handled ?

Thanks

[a comment by an automated account is not shown]

**Comment by humcqc** [comment]
Hello,

I've found a workaround using quarkus.hibernate-orm.unsupported-properties : 
```
quarkus.hibernate-orm.unsupported-properties."hibernate.multiTenancy" = DISCRIMINATOR
quarkus.hibernate-orm.unsupported-properties."hibernate.tenant_identifier_resolver" = com.digicopro.base.multitenancy.TenantIdentifierResolver
```
The new issue is that in this way my TenantIdentifierResolver that use @RequestScope does not have access to the RoutingContext.
```
import io.vertx.ext.web.RoutingContext;
import jakarta.enterprise.context.RequestScoped;
import jakarta.inject.Inject;
import org.hibernate.context.spi.CurrentTenantIdentifierResolver;
import org.jboss.logging.Logger;

@RequestScoped
public class TenantIdentifierResolver implements CurrentTenantIdentifierResolver {
    private static final Logger LOG = Logger.getLogger(TenantIdentifierResolver.class);
    @Inject
    RoutingContext context;

    @Override
    public String resolveCurrentTenantIdentifier() {
        final String tenantId = context.get(TenantEntity.TENANT_ID_PROPERTY_NAME);
        LOG.debugv("TenantId = {0}", tenantId);
        return tenantId;
    }

    @Override
    public boolean validateExistingCurrentSessions() {
        return true;
    }

}
```

and it produces : java.lang.NullPointerException: Cannot invoke "io.vertx.ext.web.RoutingContext.get(String)" because "this.context" is null.
Certainly some glue missing between quarkus CDI and hibernate config. 

Any other way to do it ? do you thing it's a missing feature that should be reported and planned ?

**Comment by humcqc** [comment]
@gsmet, @yrodiere , @Sanne do you think it could be because of [FastBootEntityManagerFactoryBuilder](https://github.com/quarkusio/quarkus/blob/3.0/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootEntityManagerFactoryBuilder.java#L192) :
```
        if (options.isMultiTenancyEnabled()) {
            options.applyCurrentTenantIdentifierResolver(new HibernateCurrentTenantIdentifierResolver(persistenceUnitName));
        }
```
and options.isMultiTenancyEnabled() just handle DATABASE and SCHEMA strategy ?

**Reply by yrodiere** [reply]
Maybe. Discriminator-based multi-tenancy is not implemented yet in Quarkus: https://github.com/quarkusio/quarkus/issues/28644 . The way to make it work would be to first add an integration test in Quarkus (under `integration-tests/hibernate-orm-tenancy`), then look at what's wrong exactly.

I suspect we'll have to look for the `@TenantId` annotation during the build to detect that we need a tenant identifier resolver even though there's no multi-tenant connection resolver (it used to always be the case when only DB multitenancy and schema multitenancy were supported).

**Reply by humcqc** [reply]
Ok, let me try and come back to you.

**Reply by humcqc** [reply]
@yrodiere , SessionFactoryOptionsBuilder from hibernate does not have anymore placeholder for multi tenancy strategy parameter. One way to do it is to re-add the multitenancy parameter that have been removed in commit : c48aceb9, do you think it's a good way ?

**Reply by yrodiere** [reply]
Reverting that commit won't work since several things changed since then, in particular the concept of multi-tenancy strategy no longer exists in Hibernate ORM, and as far as Hibernate ORM is concerned "enabling multi-tenancy" means enabling database/schema multi-tenancy, not discriminator multi-tenancy (I think there might be a bug there...).

Anyway, yes, if you detect the presence of `@TenantId` at build time and need to pass the information "there is a @TenantId", that would involve passing a boolean to the constructor of `FastBootEntityManagerFactoryBuilder`, in a way similar to what was done before that commit.

Alternatively, I just noticed you might be able to detect the presence of `@TenantId` directly in `FastBootEntityManagerFactoryBuilder`, by exploring the `metadata`. Something like:

```
        if (options.isMultiTenancyEnabled() || metadata.getFilterDefinition(TenantIdBinder.FILTER_NAME) != null) {
```

This seems simpler and more reliable... WDYT?

**Reply by humcqc** [reply]
I have a working version with the way I've proposed, but your proposal seems simpler and avoid using some deprecated class. 
Let me try and come back to you.
One question: if we use this way we don't take into account the quarkus.hibernate-orm.multitenant=discriminator property.

**Reply by yrodiere** [reply, marked as answer]
>  One question: if we use this way we don't take into account the quarkus.hibernate-orm.multitenant=discriminator property.

The multitenancy strategy (which is what this property is about) is deprecated. The concept no longer exists in Hibernate ORM.

I think eventually we'll just get rid of it and move schema/database multi-tenancy to the datasource (#11861), and will have to detect whether multi-tenancy is needed in ORM based on the datasource and the presence of `@TenantId`.

That being said, the solution I suggested works for `FastBootEntityManagerFactoryBuilder`, but you can't use it in `io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor#multitenancy`. You don't have the metadata around in that class.

So maybe for now the easiest would be the solution you already implemented. If it works, and there's integration tests, just send a PR and I'll have a look. We'll take care of removing the multitenancy strategy later.

**Reply by humcqc** [reply]
Ok, will be handled in this ticket : https://github.com/quarkusio/quarkus/issues/28644
PR: https://github.com/quarkusio/quarkus/pull/32855

**Reply by humcqc** [reply]
After this PR , I would like to discuss Hybrid mode for multitenancy. My requirements are to be able to store small tenants with Discriminator strategy, some medium tenants into different schemas, and the biggest in different databases. And this dynamically if possible. I didn't find any discussion on this , but perhaps I miss ones ?

**Reply by yrodiere** [reply]
I don't think this was discussed. Maybe open an issue to discuss that?

The first step would be to check whether it's possible in Hibernate ORM itself, without Quarkus. If it is, then I think we should be able to make it work simply by making the "multitenancy" configuration property multi-valued.
